app/src/main/java/com/materiiapps/gloom/ui/viewmodels/main/MainViewModel.kt

```diff
@@ -0,0 +1,47 @@
+package com.materiiapps.gloom.ui.viewmodels.main
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.materiiapps.gloom.api.repository.GithubAuthRepository
+import com.materiiapps.gloom.api.repository.GraphQLRepository
+import com.materiiapps.gloom.api.utils.fold
+import com.materiiapps.gloom.api.utils.ifSuccessful
+import com.materiiapps.gloom.domain.manager.AuthManager
+import com.materiiapps.gloom.ui.utils.clearRootNavigation
+import kotlinx.coroutines.launch
+
+class MainViewModel(
+    private val gqlRepository: GraphQLRepository,
+    private val authRepository: GithubAuthRepository,
+    val authManager: AuthManager
+) : ViewModel() {
+
+    fun loginWithOAuthCode(code: String, onLoggedIn: () -> Unit) {
+        viewModelScope.launch {
+            authManager.setAuthState(loading = true)
+            authRepository.getAccessToken(code).fold(
+                success = { token ->
+                    gqlRepository.getAccountInfo(token.accessToken).ifSuccessful { account ->
+                        authManager.addAccount(
+                            id = account.id,
+                            token = token.accessToken,
+                            type = authManager.awaitingAuthType!!,
+                            username = account.login,
+                            avatarUrl = account.avatarUrl,
+                            displayName = account.name,
+                            notificationCount = account.notificationListsWithThreadCount.totalCount
+                        )
+                        authManager.switchToAccount(account.id)
+                        clearRootNavigation()
+                        onLoggedIn()
+                    }
+                    authManager.setAuthState(authType = null, loading = false)
+                },
+                error = { authManager.setAuthState(authType = null, loading = false) },
+                failure = { authManager.setAuthState(authType = null, loading = false) },
+                empty = { authManager.setAuthState(authType = null, loading = false) }
```

Comment:
maybe extract these into a single function inside the viewmodel
fun AuthManager.clearAuthState(), then use like 
```kt
error = ::clearAuthState
```